### ISSUE TITLE
Discontinuity in fields for magpylib.magnet.CylinderSegment at phi-direction "internal interface"

### ISSUE THREAD

**Issue by apdioguardi** [magpylib/magpylib#512]
### What is your question?

For full cylinder segments, i.e. phi1+360=phi2, getB and getH return zeros at the "internal interface" in the phi-direction of the cylinder segment. After testing some other cases, like a half cylindrical shell, this seems to be an extension of the numerical stability near interfaces as discussed already in the physics and computation section of the docs. It makes sense then, by extension, that the behavior would be funky near an "internal interface." Maybe it would be useful to add a note to the docstring for CylinderSegment for this special case...

Here's an example:

```python
import numpy as np
import magpylib as magpy

# create a cylindrical shell
r1 = 1
r2 = 2
h = 2
phi1 = 0
phi2 = 360    # also occurs for general phi1+360 =phi2 (I tested just the phi1=90 case though)
# with magnetization
m = np.array([1,1,1])
dims = np.array([r1, r2, h, phi1, phi2])
src = magpy.magnet.CylinderSegment(magnetization=m, dimension=dims)

# one gets discontinuities (all zeros) at the boundary
print( src.getB((1.5, -1e-6, 0.5)) )
print( src.getB((1.5,   0.0, 0.5)) )
print( src.getB((1.5,  1e-6, 0.5)) )
# result:
#[0.52602734 0.8830608  0.68333994]
#[0. 0. 0.] # discontinuity produced
#[0.5260268  0.88306033 0.68334   ]

# everything looks good for the points on the opposite side or if one moves the boundary
print( src.getB((-1.5, -1e-6, 0.5)) )
print( src.getB((-1.5,   0.0, 0.5)) )
print( src.getB((-1.5,  1e-6, 0.5)) )
# result:
#[0.4335992  0.88306026 0.59091234]
#[0.43359947 0.88306056 0.59091237]
#[0.43359974 0.88306086 0.5909124 ]
```

Here is a plot to show that the fields behave nicely outside, but there is a discontinuity of zeros in the x-z plane at y=0 in the above example:

![CylinderSegment_internal_interface_example](https://user-images.githubusercontent.com/2583000/161516115-e2b3b103-8774-4d2d-8dfb-a6e9bc65dec7.PNG)

**Comment by OrtnerMichael** [comment]
hi @apdioguardi 

thanks for pointing this out. You actually found some unwanted behavior there. As you thought, the field is set to 0 close to the interface. This is not the case anymore for all computations with v4, but the CylinderSegment still makes use of this.

In the above plots you use a shell that goes from 0 to 360 ... so as a result the shell boundary is a cross- section at the 0 angle. If you would define your source from -180 to 180, then the left cross section of your stream plots would produce 0 fields.

I'll make sure to add a note about the interface to the docstrings.

If you want to do full hollow cylinders consider use "cut-outs" like in the following code (part of the doku once v4 is published), which is also much faster than the Cylinder Segment:

```python
from magpylib.magnet import Cylinder

# ring with cut-out
inner = Cylinder(magnetization=(-1,-1,-1), dimension=(4,1))
outer = Cylinder(magnetization=(1,1,1), dimension=(6,1))
ring1 = inner + outer
```

**Comment by Alexboiboi** [comment]
> If you want to do full hollow cylinders consider use "cut-outs" like in the following code (part of the doku once v4 is published), which is also much faster than the Cylinder Segment:

Is there a good reason not to catch this case internally and make the computation of a full 360° `CylinderSegment` with the cylinder core function with the cut-out method? 
Should be trivial to implement, what do you think @OrtnerMichael ?

**Comment by OrtnerMichael** [comment]
yes its trivial ... and maybe now, with the sparated core package, its also the right way.

previously it would have defeated the purpose to switch to the Cylinder core function when calling getB from CylinderSegment.

but with our new structure it makes a lot of sense.

v4 ?

**Comment by OrtnerMichael** [comment]
Problem fixed with #514
